File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/paging/NewsDataSourceFactory.java
package com.andruid.magic.newsdaily.paging;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import androidx.paging.PageKeyedDataSource;

import com.andruid.magic.newsloader.model.News;

public class NewsDataSourceFactory extends DataSource.Factory<Integer, News> {
    private MutableLiveData<PageKeyedDataSource<Integer, News>> liveDataSource =
            new MutableLiveData<>();
    private String country;

    public NewsDataSourceFactory(String country){
        this.country = country;
    }

    @NonNull
    @Override
    public DataSource<Integer, News> create() {
        NewsDataSource dataSource = new NewsDataSource(country);
        liveDataSource.postValue(dataSource);
        return dataSource;
    }

    MutableLiveData<PageKeyedDataSource<Integer, News>> getLiveDataSource() {
        return liveDataSource;
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/paging/NewsDataSourceFactory.kt
package com.andruid.magic.newsdaily.paging

import androidx.paging.DataSource
import com.andruid.magic.newsloader.model.News
import kotlinx.coroutines.CoroutineScope

class NewsDataSourceFactory(private val scope: CoroutineScope,
                            private val country: String,
                            private val category: String
) : DataSource.Factory<Int, News>() {
    override fun create(): DataSource<Int, News> {
        return NewsDataSource(scope, country, category)
    }
}